### PR TITLE
[monitor][txn] Add metrics for transaction

### PR DESCRIPTION
**Pull request by tjiuming**
### Motivation

add metrics for transaction

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [x] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by tjiuming** [comment]
Tests to be completed

**Comment by Anonymitaet** [comment]
@momo-jun a soft reminder: here is a PR w/ doc-required label, could u pls follow up? Thanks

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java:45]: because topic transaction stats need transaction buffer stats and pending ack stats, so I think it only represent tb ongoingTxnCount

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java:136]: pulsar_txn_ tb_active_count

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java:138]: pulsar_txn_ tb_aborted_count

Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java:138]: should add pulsar_txn_tb_committed_count

**Review by congbobo184 — COMMENTED** [review]
Good work! as the seem as tb, we should add tp stats

**Comment by tjiuming** [comment]
> Good work! as the seem as tb, we should add tp stats

I know TC means `TransactionCoordinator` and TB means `TransactionBuffer`, but what means TP?

**Comment by tjiuming** [comment]
> > Good work! as the seem as tb, we should add tp stats
> 
> I know TC means `TransactionCoordinator` and TB means `TransactionBuffer`, but what means TP?

Does it mean `TransactionPendingAckStore` ?

**Comment by tjiuming** [comment]
> > Good work! as the seem as tb, we should add tp stats
> 
> I know TC means `TransactionCoordinator` and TB means `TransactionBuffer`, but what means TP?

Does it mean `TransactionPendingAckStore` ?

**Comment by momo-jun** [comment]
Hi @tjiuming will you submit a doc PR as soon as this PR is approved and merged? Ping me if you need any review or other assistance.

**Comment by tjiuming** [comment]
> 

I‘ll create another PR to add doc after this PR finished

**Comment by congbobo184** [comment]
yes PendingAck is TP

**Comment by tjiuming** [comment]
@codelipenghui  @gaoran10 Could you please help review?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleStatsImpl.java
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleStatsImpl.java:29]: Need to check if topic level metrics is enabled, otherwise we should only expose the namespace level metrics

**Review by codelipenghui — CHANGES_REQUESTED** [review]
Great change, left a comment about the transaction pending ack stats.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleStatsImpl.java
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleStatsImpl.java]: fixed

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientTest.java
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientTest.java:206]: should add the request to pending and check  the value > 0D is better

**Review by congbobo184 — COMMENTED** [review]
LGTM

**Review by congbobo184 — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java
Comment (gaoran10) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:26]: @congbobo184 @codelipenghui  Do we need to distinguish between commit(abort) TB or commit(abort) PendingAck? Same as operation latency.

Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:26]: PendingAck is in [PendingAckHandleStats.java](https://github.com/apache/pulsar/pull/15140/files#diff-13ac31e4202c3d4048fcbac33099531393cf361e9e0ebbd74ef8680ea18b2708)

Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:26]: yes, we need to distinguish TB and TP is better @tjiuming

Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:26]: @congbobo184 @congbobo184  Could you please describe it more clearly? I mean, there were already distinguished between TB and TP, please see: [TransactionBuffer.java](https://github.com/apache/pulsar/pull/15140/files#diff-819cb066b87a312a85c161d7bccf5699ddd8e1aded6545a6f39a237d16e6e707) and [PendingAckHandleStatsImpl](https://github.com/apache/pulsar/pull/15140/files#diff-3e4c8d1822aa122c6b58f1b2c8227a6dac1354cbeeb1f68cab24a6587a881ea5)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1870]: nit: can we call getTransactionBuffer() only once ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientImpl.java
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientImpl.java:62]: I think that we usually use "nanos", because millis will be usually 0 (hopefully)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java:52]: are we using the default global registry ?
what happens if you run two brokers inside the same JVM ?
is it a pattern that we follow for other metrics ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleStatsImpl.java
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleStatsImpl.java:41]: do we have other metrics that report "unknown" as topic when it is not set ?
I think that we should throw a error in this case, it is not possible to see a blank topic name here

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleStatsImpl.java
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleStatsImpl.java:41]: right, it's not possible that topic is blank, it just in case of unexpected behaviors.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java:52]: Yes. 
When run 2 brokers inside a same JVM, it just bind 2 different ports, other behaviors makes no differences.
Yes, many other metrics using the default registry.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1870]: it makes sense, I‘ll change it.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java
Comment (nicoloboschi) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:252]: IIUC `ongoingTxns` and `aborts` are real gauges because they can be increased or decreased. 
`txnCommittedCounter` is only incremented and so it's a counter. 
Is it correct ? 
If so, I believe we have to introduce a txnAbortedCounter

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java
Comment (nicoloboschi) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:34]: we can use `AutoCloseable`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java
Comment (nicoloboschi) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java:114]: log the exception ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java
Comment (nicoloboschi) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java:121]: why `instance == this` is needed ? if I call close() on this instance I expect to call this object even if it's not the singleton

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/PendingAckHandleStats.java
Comment (nicoloboschi) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/PendingAckHandleStats.java:29]: Autocloseable?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java:121]: it makes sense.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:252]: it makes sense.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferClientStatsImpl.java:114]: I believe no need, it just in case of TopicName.get() throws exception.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:26]: I see [TransactionBuffer.java](https://github.com/apache/pulsar/pull/15140/files#diff-819cb066b87a312a85c161d7bccf5699ddd8e1aded6545a6f39a237d16e6e707) and [PendingAckHandleStatsImpl](https://github.com/apache/pulsar/pull/15140/files#diff-3e4c8d1822aa122c6b58f1b2c8227a6dac1354cbeeb1f68cab24a6587a881ea5) 
The implementation is different; what is the consideration here? I missed a part of the code before, sorry

```
    public void recordCommitTxn(boolean success) {
        commitTxnCounter.labels(success ? labelSucceed : labelFailed).inc();
    }
```

```
    private final Counter abortFailed;
    private final Counter commitFailed;
    private final Summary abortLatency;
    private final Summary commitLatency;
```

I think in TransactionBuffer the latency is managedLedger append entry latency, it also already exists. So we only need to add the count and the rate.the same is true for PendingAckStats

Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:26]: For TransactionBuffer, I recorded TransactionBufferClient's latency, because there is RPC request existing.

> The implementation is different; what is the consideration here

In TransactionBufferClientStats, abort/commit latency only recorded when abort/commit success, so we can get `abortSuccess` and `commitSuccess` from `abortLatency` `commitLatency `.
@congbobo184

Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:26]: If TransactionBuffer add the latency, I think PendingAck also need to add this metrics

Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferClientStats.java:26]: `pulsar_txn_tp_commit_latency` added

**Comment by tjiuming** [comment]
@codelipenghui @congbobo184 This PR has been blocked for a long while, please help review and merge ASAP

**Review by gaoran10 — APPROVED** [review]

**Review by Shoothzj — CHANGES_REQUESTED** [review]
new metrics should ends with "_total"

See https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md

Also see
- https://github.com/apache/pulsar/pull/13785
- https://github.com/apache/pulsar/pull/16591
- https://github.com/apache/pulsar/pull/16610
- https://github.com/apache/pulsar/pull/16611

**Comment by tjiuming** [comment]
@Shoothzj  PTAL

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/PendingAckPersistentTest.java
Comment (Shoothzj) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/PendingAckPersistentTest.java:252]: this's should be total too?

Comment (tjiuming) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/PendingAckPersistentTest.java]: yes, sorry, I forgot it.

**Review by Shoothzj — APPROVED** [review]

**Pull request merged**
